py/desiutil/timer.py

```diff
@@ -0,0 +1,125 @@
+"""
+timer: standardizing reporting of algorithm and I/O timing
+
+The `timer.Timer` class is intended for reporting timing of events that
+take seconds to minutes; it is not intended for detailed profiling.  Example::
+
+    from desiutil.timer import Timer
+    t = Timer()
+    t.start('blat')
+
+    #- Use context manager for timing simple steps, e.g. one-liners
+    with t.time('blat.input'):
+        stuff = io.read(filename)
+
+    #- Or use full start/stop if you prefer
+    t.start('blat.algorithm')
+    results = calculate(stuff)
+    t.stop('blat.algorithm')
+
+    with t.time('blat.output'):
+        io.write(results)
+
+    #- Stop outer timer; Note that named timers can be nested or interleaved
+    t.stop('blat')
+
+    #- Print a json report of the timing
+    print(t.report())
+"""
+
+import traceback
+import time
+import datetime
+import json
+import copy
+import os.path
+from contextlib import contextmanager
+
+class Timer(object):
+    """
+    A basic timer class for standardizing reporting of algorithm and I/O timing
+    
+    TIMER:<START|STOP>:<filename>:<lineno>:<funcname>: <message>
+    """
+    
+    def __init__(self):
+        self.timers = dict()
+    
+    def _prefix(self, step):
+        """
+        Return standard prefix string for timer reporting
+        
+        Args:
+            step (str): timing step, e.g. "START" or "STOP"
+        """
+        stack = traceback.extract_stack()
+        #- Walk backwards in stack to find first caller not from this file
+        #- and not contextlib.py
+        #- (exact index depends on whether context manager was used or not)
+        thisfile = os.path.normpath(__file__)
+        for caller in stack[-1::-1]:
+            if os.path.normpath(caller.filename) != thisfile and \
+               os.path.basename(caller.filename) != 'contextlib.py':
+                break
+        
+        filename = os.path.basename(caller.filename)
+        return f"TIMER-{step}:{filename}:{caller.lineno}:{caller.name}:"
+    
+    def start(self, name):
+        """Start a timer `name` (str)"""
+        # timestamp = datetime.datetime.now().replace(microsecond=0).isoformat()        
+        timestamp = datetime.datetime.now().isoformat()        
+        if name in self.timers:
+            print(self._prefix('WARNING'), f'Restarting {name} at {timestamp}')
+        
+        print(self._prefix('START'), f'Starting {name} at {timestamp}')
+        self.timers[name] = dict(start=time.time())
+    
+    def stop(self, name):
+        """Stop timer `name` (str)"""
```

Comment:
Document that `stop()` returns a value.  See also next comment.

Reply:
done